### ISSUE TITLE
Unlink 16-bit Precision and AMP

### ISSUE THREAD

**Issue by tsteffek** [PyTorchLightning/pytorch-lightning#3556]
## 🚀 Feature
The [Trainer's precision argument](https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#precision), currently allows for 16- or 32-bit precision. However, 16-bit precision does not actually set the model to 16-bit precision, it instead enables [PyTorch's Autocast](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast).

I'm proposing unlinking the two and providing a third option, 'auto'/'autocast' or 'amp'.

### Motivation

1. Hiding autocast behind 16-bit precision doesn't seems intuitive. I ask for 16-bit, I expect 16-bit. Receiving a float32 tensor from your softmax layer when you expected 16-bit might even case issues for some people.
2. While autocast makes sane assumptions about which option should be cast to what precision, it does not grant full control to the user. Maybe I need to run absolutely everything in 16-bit to fit the model into RAM.
3. [AdaptiveLogSoftmaxWithLoss](https://pytorch.org/docs/master/generated/torch.nn.AdaptiveLogSoftmaxWithLoss.html) is currently [broken within autocast](https://github.com/pytorch/pytorch/issues/44724). The option to switch to 16-bit precision instead of autocast would let me avoid this issue, while still providing 16-bit precision to the rest of my model.

### Pitch

`Trainer(precision=16)` should transform (all) models and inputs to float16.
Autocast should be available through a separate option, e.g. `Trainer(precision='auto')`.

### Alternatives

I've considered an additional flag for trainer, such as `Trainer(autocast=True)`, however as PyTorch urges you to not call `.half()` on your models and inputs before entering the autocast region, this seems to clash with any input to the precision argument.

### Additional context

In 7-10 days I might have enough time on my hands to take a shot at the PR myself. A first guess would be to follow the code that moves models and inputs to different devices, as both utilize the `.to()` method. Let me know if you're interested and if you see any problem with that approach.

**Comment by s-rog** [comment]
`with torch.cuda.amp.autocast(enabled=False):`

For now, as a workaround I used this in combination with `precision=16` (for some fourier transform stuff)

**Comment by ruotianluo** [comment]
I believe 
```
        precision=16,
        amp_backend='apex',
        amp_level='O2',
```
gives you precision 16.

**Comment by tsteffek** [comment]
> `with torch.cuda.amp.autocast(enabled=False):`
> 
> For now, as a workaround I used this in combination with `precision=16` (for some fourier transform stuff)

Does this really give you 16-bit precision? When I took a brief look into the relevant code, I didn't see anything that changes to 16-bit aside from autocast. So as to my understanding, that shouldn't do anything?

**Comment by tsteffek** [comment]
So I never got a real go-ahead with this one, is this and/or my contribution wanted?
